src/Swarm/Game/WorldGen.hs

```diff
@@ -66,46 +70,58 @@ testWorld2Entites =
     , "copper ore"
     ]
 
--- | A more featureful test world.
-testWorld2 :: Seed -> WorldFun TerrainType Text
-testWorld2 baseSeed (Coords ix@(r, c)) =
+-- | Look up an entity name in an entity map, when we know the entity
+--   must exist.  This is only used for entities which are named in
+--   'testWorld2'.
+readEntity :: EntityMap -> Text -> Entity
+readEntity em name =
+  fromMaybe
+    (error $ "Unknown entity name in WorldGen: " <> show name)
+    (lookupEntityName name em)
+
+-- | The main world of the classic game, for historical reasons named
+--   'testWorld2'.  If new entities are added, you SHOULD ALSO UPDATE
+--   'testWorld2Entities'.
+testWorld2 :: EntityMap -> Seed -> WorldFun TerrainType Entity
+testWorld2 em baseSeed = second (readEntity em) . WF $ \(Coords ix) ->
   genBiome
+    ix
     (bool Small Big (sample ix pn0 > 0))
     (bool Soft Hard (sample ix pn1 > 0))
     (bool Natural Artificial (sample ix pn2 > 0))
  where
-  h = murmur3 0 (into (show ix))
+  h ix = murmur3 0 (into (show ix))
```

Comment:
Why is it necessary to put `ix` everywhere? `H` never seems to use anything else. :raised_eyebrow:

Reply:
Since I made `WorldFun` into a newtype, `ix` is no longer in scope.  We used to be able to write  something like `testWorld2 baseSeed ix@(Coords ...) = ...` but now we have to write  `testWorld2 baseSeed = WF $ \ix@(Coords ...) -> ...` .   I think it's worth it because newtypes are good documentation, and we get a convenient `Bifunctor` instance, but it is kind of annoying.  Unless there's a better way to refactor the code now that I haven't thought of.

Reply:
OK, never mind, I fixed it with some nested `where` blocks.